library/Freckle/App/Kafka/Consumer.hs

```diff
@@ -0,0 +1,166 @@
+
+readKafkaTopics :: String -> Either String (NonEmpty TopicName)
+readKafkaTopics t = case NE.nonEmpty $ T.splitOn "," $ T.pack t of
+  Just xs@(x NE.:| _)
+    | x /= "" -> Right $ TopicName <$> xs
+  _ -> Left "Kafka topics cannot be empty"
+
+envKafkaOffsetReset
+  :: Env.Parser Env.Error OffsetReset
+envKafkaOffsetReset =
+  Env.var
+    (eitherReader readKafkaOffsetReset)
+    "KAFKA_OFFSET_RESET"
+    $ Env.def Earliest
+
+readKafkaOffsetReset :: String -> Either String OffsetReset
+readKafkaOffsetReset t = case T.pack t of
+  "earliest" -> Right Earliest
+  "latest" -> Right Latest
+  _ -> Left "Kafka offset reset must be one of earliest or latest"
+
+envKafkaConsumerConfig
+  :: Env.Parser Env.Error KafkaConsumerConfig
+envKafkaConsumerConfig = do
+  brokerAddresses <- envKafkaBrokerAddresses
+  consumerGroupId <- Env.var Env.nonempty "KAFKA_CONSUMER_GROUP_ID" mempty
+  kafkaTopics <- envKafkaTopics
+  kafkaOffsetReset <- envKafkaOffsetReset
+  kafkaExtraProps <-
+    Env.var
+      (fmap Map.fromList . keyValues)
+      "KAFKA_EXTRA_SUBSCRIPTION_PROPS"
+      (Env.def mempty)
+  pure $
+    KafkaConsumerConfig
+      brokerAddresses
+      consumerGroupId
+      kafkaTopics
+      kafkaOffsetReset
+      kafkaExtraProps
+
+class HasKafkaConsumer env where
+  kafkaConsumerL :: Lens' env KafkaConsumer
+
+consumerProps :: KafkaConsumerConfig -> ConsumerProperties
+consumerProps KafkaConsumerConfig {..} =
+  brokersList brokers
+    <> groupId kafkaConsumerConfigGroupId
+    <> noAutoCommit
+    <> logLevel KafkaLogInfo
+ where
+  brokers = NE.toList kafkaConsumerConfigBrokerAddresses
+
+subscription :: KafkaConsumerConfig -> Subscription
+subscription KafkaConsumerConfig {..} =
+  topics (NE.toList kafkaConsumerConfigTopics)
+    <> offsetReset kafkaConsumerConfigOffsetReset
+    <> extraSubscriptionProps kafkaConsumerConfigExtraSubscriptionProps
+
+withKafkaConsumer
+  :: MonadUnliftIO m
+  => KafkaConsumerConfig
+  -> (KafkaConsumer -> m a)
+  -> m a
+withKafkaConsumer config = bracket newConsumer closeConsumer
+ where
+  (props, sub) = (consumerProps &&& subscription) config
+  newConsumer = either throwIO pure =<< Kafka.newConsumer props sub
+  closeConsumer = maybe (pure ()) throwIO <=< Kafka.closeConsumer
+
+timeoutMs :: Timeout -> Int
+timeoutMs = \case
+  TimeoutSeconds s -> s * 1000
+  TimeoutMilliseconds ms -> ms
+
+runConsumer
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , MonadLogger m
+     , HasKafkaConsumer env
+     , FromJSON a
+     )
+  => Timeout
+  -> (a -> m ())
+  -> m ()
+runConsumer pollTimeout onMessage = void $ Immortal.create $ \thread -> Immortal.onUnexpectedFinish thread handleException $ do
+  consumer <- view kafkaConsumerL
+  eMessage <-
+    pollMessage consumer $ Kafka.Timeout $ timeoutMs pollTimeout
+  case eMessage of
+    Left (KafkaResponseError RdKafkaRespErrTimedOut) -> logInfo $ "Polling timeout"
+    Left err -> logError $ "Error polling for message from Kafka" :# ["error" .= show err]
```

Comment:
Any reason in particular to split these 2 errors up?

Reply:
The polling timeout error is expected, frequently. Every time there is no event withing some period it will throw that. I'd probably DEBUG, FWIW.

Reply:
Aha! Yeah I think debug is the way to go then